janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/JanusGraphMixedIndexAggStep.java

```diff
@@ -25,54 +25,86 @@
+
+import static org.janusgraph.graphdb.database.util.IndexRecordUtil.key2Field;
 
 /**
  * A custom count step similar to {@link org.apache.tinkerpop.gremlin.process.traversal.step.map.CountGlobalStep} but
  * uses mixed index query to directly fetch number of satisfying elements without actually fetching the elements.
  *
  * @author Boxuan Li (liboxuan@connect.hku.hk)
  */
-public class JanusGraphMixedIndexCountStep<S> extends ReducingBarrierStep<S, Long> implements Profiling {
+public class JanusGraphMixedIndexAggStep<S> extends ReducingBarrierStep<S, Number> implements Profiling {
 
     private final ArrayList<HasContainer> hasContainers = new ArrayList<>();
-    private MixedIndexCountQuery mixedIndexCountQuery = null;
+    private MixedIndexAggQuery mixedIndexAggQuery = null;
     private boolean done;
+    private final Aggregation aggregation;
 
-    public JanusGraphMixedIndexCountStep(JanusGraphStep janusGraphStep, Traversal.Admin<?, ?> traversal) {
+    public JanusGraphMixedIndexAggStep(JanusGraphStep janusGraphStep, Traversal.Admin<?, ?> traversal, Aggregation agg) {
         super(traversal);
         JanusGraphTransaction tx = JanusGraphTraversalUtil.getTx(traversal);
-
-        final MixedIndexCountQueryBuilder countQueryBuilder = (MixedIndexCountQueryBuilder) tx.mixedIndexCountQuery();
+        aggregation = agg;
+        final MixedIndexAggQueryBuilder aggregationQueryBuilder = (MixedIndexAggQueryBuilder) tx.mixedIndexAggQuery();
 
         final GraphCentricQuery query = janusGraphStep.buildGlobalGraphCentricQuery();
 
         if (query != null && query.getIndexQuery().isFitted()) {
-            final JointIndexQuery indexQuery = query.getIndexQuery().getBackendQuery();
-            mixedIndexCountQuery = countQueryBuilder.constructIndex(indexQuery,
-                Vertex.class.isAssignableFrom(janusGraphStep.getReturnClass()) ? ElementCategory.VERTEX : ElementCategory.EDGE);
+            final String fieldName = agg.getFieldName();
+            boolean isIndexed;
+            if (fieldName == null) {
+                isIndexed = true;
+            } else {
+                MixedIndexType indexType = (MixedIndexType)query.getIndexQuery().getBackendQuery().getQuery(0).getIndex();
+                Optional<? extends Class<?>> dataType = Arrays.stream(indexType.getFieldKeys())
+                    .filter(f -> f.getFieldKey().name().equals(fieldName))
+                    .map(f -> f.getFieldKey().dataType())
+                    .filter(Number.class::isAssignableFrom)
+                    .findFirst();
+                if (dataType.isPresent()) {
+                    isIndexed = true;
+                    aggregation.setDataType(dataType.get());
+                    RelationType rt = tx.getRelationType(fieldName);
+                    if (rt instanceof PropertyKey)
+                        aggregation.setFieldName(key2Field(indexType, (PropertyKey) rt));
+                } else {
+                    isIndexed = false;
```

Comment:
Maybe I am missing something, but is this branch covered by any test case? I didn't find any but not sure why codecov didn't complain.